### ISSUE TITLE
Cannot load KERAS LSTM model if they contain KERAS Masking Layer

### ISSUE THREAD

**Issue by mzattera** [deeplearning4j/deeplearning4j#10166]
#### Issue Description

I am using [KNIME](https://www.knime.com/) to train KERAS models that I then import into deeplearning4j. Recently, I have been using  a simple LSTM network (64 LSTM units connected to a softmax fully connected layer). As in KNIME KERAS nodes the only way of training the model with sequence of different length is to add a KERAS masking layer, I added it to my model, but as soon as I did it, it stopped loading in deeplearning4j with java.lang.ClassCastException (see below). The model loads if I do not use the masking layer.

#### Version Information

I have the below pom.xml configuration for deeplearnign4j. I am using KNIME 4.7.8 to generate the KERAS NN

	  
	  <dependency>
	      <groupId>org.deeplearning4j</groupId>
	      <artifactId>deeplearning4j-core</artifactId>
	      <version>1.0.0-M2.1</version>
	  </dependency>
        
		
		<dependency>
		    <groupId>org.nd4j</groupId>
		    <artifactId>nd4j-cuda-11.6</artifactId>
		    <version>1.0.0-M2.1</version>
		</dependency>
		<dependency>
		    <groupId>org.nd4j</groupId>
		    <artifactId>nd4j-cuda-11.6</artifactId>
		    <version>1.0.0-M2.1</version>
		    <classifier>windows-x86_64-cudnn</classifier>
		</dependency>  
		<dependency>
		    <groupId>org.bytedeco</groupId>
		    <artifactId>cuda-platform-redist</artifactId>
		    <version>11.6-8.3-1.5.7</version>
		</dependency>

#### Additional Information

Exception Stacktrace:

```
Exception in thread "main" java.lang.ClassCastException: class org.deeplearning4j.nn.conf.layers.recurrent.LastTimeStep cannot be cast to class org.deeplearning4j.nn.conf.layers.FeedForwardLayer (org.deeplearning4j.nn.conf.layers.recurrent.LastTimeStep and org.deeplearning4j.nn.conf.layers.FeedForwardLayer are in unnamed module of loader 'app')
	at org.deeplearning4j.nn.modelimport.keras.layers.recurrent.KerasLSTM.setWeights(KerasLSTM.java:452)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelUtils.importWeights(KerasModelUtils.java:352)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:188)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:97)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildModel(KerasModelBuilder.java:311)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:150)
```

Code I use to load the model:

`ComputationGraph kerasModel = KerasModelImport.importKerasModelAndWeights("kerasModel.h5", false);`

The .h5 file I use is attached: [kerasModel.zip](https://github.com/user-attachments/files/18259622/kerasModel.zip)

**Comment by agibsonccc** [comment]
Odd..we have so many test cases for LSTMs. Do you know if this is using keras 3?

**Comment by mzattera** [comment]
I don't know the version KNIME uses, sorry :( but I attached the .h5 file; could you give it a quick try? Jut loading it and see if it works on your machine? Maybe I am doing something wrong. I am used to load models created with this same version of KNIME, so it should work. I think it has to do with masking but, again, I might be wrong.

**Comment by agibsonccc** [comment]
@mzattera what knime version are you using? I'm sure they'll tell you what version they're using from that. I can at least try to find it on my end. Given it's EOY and this isnt paid work I"m unfortunately going to have to wait till after the new year holiday to debug this in depth.

**Comment by mzattera** [comment]
I am using KNIME 4.7.8 in the meantime, I can check if I find the Keras version they are using.

**Comment by agibsonccc** [comment]
@mzattera  https://docs.knime.com/2018-12/deep_learning_installation_guide/index.html#known-issues it looks like it's keras 2.x. We should be able to load that then...let me double check on my end.

**Comment by mzattera** [comment]
> Given it's EOY and this isn't paid work I'm unfortunately going to have to wait till after the new year holiday to debug this in depth.

I'll be happy to give a contribution for the fix if you can kindly tell me what will be the easiest way of doing it, regardless when this is going to be fixed: it is not urgent and can be done next year anyway.

**Comment by agibsonccc** [comment]
@mzattera  it might already be fixed in the current master. I need to test against M2.1 and SNAPSHOT. Unfortunately that's only available in the repo atm due to a major rewrite of the internals  going on. I'd need to check more on what's going on with knime though. Are they using M2.1? To make my job easier, maybe try importing the model manually in java first.

**Comment by mzattera** [comment]
> To make my job easier, maybe try importing the model manually in java first.

Sorry, I am new to deeplearning4j, not sure what you mean with this. Only way I know to load a motel is the 1-liner snippet I put when I opened the issue :(

BTW, looking forward for the new release; I do like this framework.

**Comment by agibsonccc** [comment]
Yeah that's actually all I need! If you can just check with m2.1 that would
be great.  If not I will check later.

**Comment by mzattera** [comment]
To be clear, I ran this line of code, with the .h5 model using M2.1 (see pom.xml above) and I get the Exception I have shown above....

`ComputationGraph kerasModel = KerasModelImport.importKerasModelAndWeights("kerasModel.h5", false);`

**Comment by agibsonccc** [comment]
Closing — this has been addressed in the upcoming release (branch ag_new_release_updates_2). The relevant code has been rewritten/fixed.
